src/main/webapp/WEB-INF/pages/productDetails.jsp

```diff
@@ -25,11 +41,41 @@
             <td>Price</td>
             <td>
                 <a href="${pageContext.servletContext.contextPath}/products/${product.id}?priceHistory=true">
-                <fmt:formatNumber value="${product.price}" type="currency" currencySymbol="${product.currency.symbol}"/>
+                    <fmt:formatNumber value="${product.price}" type="currency"
+                                      currencySymbol="${product.currency.symbol}"/>
                 </a>
             </td>
         </tr>
     </table>
     <br>
+    <form method="post">
+        <p>Quantity: </p>
+        <input name="quantity" value="${not empty error ? param.quantity : 1}">
+        <button>Add to cart</button>
+        <c:if test="${not empty error}">
+            <div style="color: red">
+                    ${error}
+            </div>
+        </c:if>
+    </form>
+
+    <br>
+    <h3>Recently viewed products:</h3>
+    <table style="table-layout: fixed;border-collapse: collapse;text-align: center;">
```

Comment:
Here and in other places: please move inline styles to CSS file to avoid code duplication, and for better code organization and performance.